### ISSUE TITLE
Unable to acquire token silently after B2C login and redirection to SPA redirectUri

### ISSUE THREAD

**Issue by danielvoigt** [AzureAD/microsoft-authentication-library-for-js#4935]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.24.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

My understanding is that the B2C flow with MSAL.js for a sign-in only policy works in the following way (please correct if mistaken):

1) create `msalInstance` with the msal config
2) call `msalInstance.handleRedirectPromise()`
3) when the `handleRedirectPromise` promise returns, call `loginRedirect` with the correct scopes, which will redirect the user to a b2c login page, and on completion, navigate back to the SPA `redirectUri`
4) once the the b2c login is complete, and the user is navigated back to the SPA `redirectUri`, we use the `msalInstance.acquireTokenSilent` with the correct scopes and current active account passed in as function option params, so we can get all the user information such as the `accessToken`. This is how we translate the information the MSAL stores in the browser (`sessionStorage` in this case) to user information and an access token that I can use. 

Steps 1-3 are working as expected. The issue is specifically that on step 4, after the redirect happens when I login, I am unable to acquire the token. There are no active accounts (when I attempt to retrieve with MSAL.js), and I am also unable to acquire the token with `acquireTokenSilent`. The error I am getting is the following:
`
BrowserAuthError: no_account_error: No account object provided to acquireTokenSilent and no active account has been set. Please call setActiveAccount or provide an account on the request.
`

The issue is that I can't call `setActiveAccount` when the active account does not exist (nothing exists for `msalInstance.getAllAccounts()`.

I am using the following code for step 4 - which is run after the the successful login redirect to the `redirectUri`:
```javascript
        const accounts = msalInstance.getAllAccounts();
        if (accounts?.length > 0) {
         msalInstance.setActiveAccount(accounts[0]);
        }

        if (msalInstance.getActiveAccount()) {
         const msalScopes = { scopes: ["my-example-b2c-web-api-scope"] }
         msalInstance
            .acquireTokenSilent(msalScopes)
            .then((tokenResponse) => {
              // Do something with the tokenResponse
              console.log('acquireTokenSilent res', tokenResponse);
            })
            .catch((error) => {
              console.log('acquireTokenSilent error', error);
              if (error instanceof msal.InteractionRequiredAuthError) {
                // fallback to interaction when silent call fails
                return msalInstance.acquireTokenRedirect(
                  msalScopes
                );
              }
            });
```

The issue is that because `this.msalInstance.getAllAccounts()` is `null` (empty), I cannot successfully retrieve the token information from the `acquireTokenSilent` call.

*However*, if I sign-in the first time, and get navigated back to the `redirectUri`, and then I run the sign-in again (the `msalInstance.handleRedirectPromise()` and then `loginRedirect()` code), I will successfully get the accounts and the token response from `acquireTokenSilent`. So I know that it most likely is not a configuration issue, or something to do with scopes or permissions.

The issue that needs to be figured out, is how can the token be retrieved on the successful navigation from the b2c login to the SPA `redirectUri`? Having to trigger a sign-in twice is not doable for the app use-case. I am hoping there is a better way to do this and to fix the issue of being able to retrieve the token on that initial redirect navigation to the `redirectUri` of the SPA.

The expected behavior is that when the SPA `redirectUri` is hit, that I will be able to get the access token and other user information. `acquireTokenSilent` doesn't seem to be working as expected for this use case.

### Error Message

`
BrowserAuthError: no_account_error: No account object provided to acquireTokenSilent and no active account has been set. Please call setActiveAccount or provide an account on the request.
`

### Msal Logs

n/a

### MSAL Configuration

```javascript
        const msalConfig = {
          auth: {
            clientId: 'example123',
            authority: 'b2c-example',
            redirectUri: 'localhost:4200/redirect-uri',
            knownAuthorities: ['my-b2c-web-api-scope'],
            postLogoutRedirectUri: 'localhost:4200/redirect-uri',
            navigateToLoginRequestUrl: true,
          },
          cache: {
            cacheLocation: 'sessionStorage',
            storeAuthStateInCookie: true,
          },
          system: {
            loggerOptions: {
              loggerCallback: (level, message, containsPii) => {
                if (containsPii) {
                  return;
                }
                switch (level) {
                  case msal.LogLevel.Error:
                    console.error(message);
                    return;
                  case msal.LogLevel.Info:
                    console.info(message);
                    return;
                  case msal.LogLevel.Verbose:
                    console.debug(message);
                    return;
                  case msal.LogLevel.Warning:
                    console.warn(message);
                    return;
                }
              },
            },
          },
        };
```

### Relevant Code Snippets

```javascript
     // initial code 
    msalInstance
      .handleRedirectPromise()
      .then((tokenResponse) => {
        if (tokenResponse === null) {
          // If the tokenResponse === null, you are not coming back from an auth redirect.
          msalInstance.loginRedirect(this.msalApiScopes);
        } else {
          // If the tokenResponse !== null, then you are coming back from a successful authentication redirect.
          console.log('success res', tokenResponse);
        }
      })
      .catch((err) => {
        // handle error, either in the library or coming back from the server
        console.log('err', err);
      });

        // code that runs after the redirectUri returns
        const accounts = msalInstance.getAllAccounts();
        if (accounts?.length > 0) {
          msalInstance.setActiveAccount(accounts[0]);
        }

         const msalScopes = { scopes: ["my-example-b2c-web-api-scope"] }
         msalInstance
            .acquireTokenSilent(msalScopes)
            .then((tokenResponse) => {
              // Do something with the tokenResponse
              console.log('acquireTokenSilent res', tokenResponse);
            })
            .catch((error) => {
              console.log('acquireTokenSilent error', error);
              if (error instanceof msal.InteractionRequiredAuthError) {
                // fallback to interaction when silent call fails
                return msalInstance.acquireTokenRedirect(
                  msalScopes
                );
            });


```

### Reproduction Steps

see initial description

### Expected Behavior

Be able to get token through `acquireTokenSilent` method directly after the redirect back to the SPA `redirectUri` after the b2c login is completed. Not having to go through the `handleRedirectPromise` a second time

### Identity Provider

Azure B2C Basic Policy

### Browsers Affected (Select all that apply)

Chrome

### Source

External (Customer)

**Comment by danielvoigt** [comment]
@jasonnutter and @hectormmg are you available to look at this? It looks you like you are actively responding to issues in this repo

**Comment by sunilshrestha** [comment]
Hi @danielvoigt did you tried using msal interceptor (https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/msal-interceptor.md). It internally handles acquireTokenSilent logic based on your configuration if you don't have any custom logic implementation.

**Comment by bmahall** [comment]
@danielvoigt I am currently looking into this , thanks for your patience!

**Comment by danielvoigt** [comment]
> Hi @danielvoigt did you tried using msal interceptor (https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/msal-interceptor.md). It internally handles acquireTokenSilent logic based on your configuration if you don't have any custom logic implementation. 

Thanks for the suggestion @sunilshrestha. I'm currently using MSAL.js v2 for the browser (@azure/msal-browser), rather than the angular specific wrapper library. However I will attempt to re-implement this with the wrapper library if I'm unable to figure out the issue with the general MSAL browser lib.

**Comment by bmahall** [comment]
@danielvoigt  Can you kindly refer to the below attached code snippet and check if it works:-
`msalInstance.handleRedirectPromise()

    .then((tokenResponse) => {
        if (!tokenResponse) {
            const accounts = msalInstance.getAllAccounts();
            if (accounts.length === 0) {
                // No user signed in
                msalInstance.loginRedirect();
            }
        } else {
            // Do something with the tokenResponse
        }
    })
    .catch(err => {
        // Handle error
        console.error(err);
    });`

Please note:   If you are calling loginRedirect or acquireTokenRedirect from a page that is not your redirectUri you will need to ensure handleRedirectPromise is called and awaited on both the redirectUri page as well as the page that you initiated the redirect from. This is because the redirectUri page will initiate a redirect back to the page that originally invoked loginRedirect and that page will process the token response.

**Comment by danielvoigt** [comment]
@bmahall The solution proposed does not work. I moved on to `msal-angular`, hoping to find a bug free solution, but I'm running into new bugs with that library as well.

**Comment by bmahall** [comment]
@danielvoigt Can you please check:-
1. Value of the redirectUri and postRedirectUri should be in valid format, scope should be passed as an array. 
Please carefully verify the value of the scope parameter being passed. 

2. Get the account from token response as below: 
     let accountObj;
      if (tokenResponse) {
        accountObj = tokenResponse.account;
      } else {
        accountObj = this.app.getAllAccounts()[0];
      }


3. When issuing a call to acquireTokenSilent() pass the active account along with the scope. 

If none of this helps, please share verbose logs and output of : console.log('success res', tokenResponse) on my email mentioned in my profile on Github. 


You are welcome to open a new issue for issues in msal-angular library. 

Thanks for your patience.

**Comment by danielvoigt** [comment]
@bmahall 
1. `redirect` and `postRedirectUri` are both valid strings. The scope is a valid array. The configuration is included in the issue description
2. As discussed in the description, there is no active account after the redirect, which is the main issue: 
> "There are no active accounts (when I attempt to retrieve with MSAL.js), and I am also unable to acquire the token with acquireTokenSilent"
3. there are no active accounts to pass along with the scope

**Comment by bmahall** [comment]
@danielvoigt After looking into the source code you have linked above, it is possible that handleRedirectPromise has not finished when you call getAllAccounts(). 

Can you please ensure that handleRedirectPromise() completes, and then use getAllAccounts() API. 
You could call getAllAccounts() inside the else block here:
`else {

          // If the tokenResponse !== null, then you are coming back from a successful authentication redirect.
          console.log('success res', tokenResponse);
          // SET ACTIVE ACCOUNT HERE
        }`
        
Let me know if this helps.

**Comment by BadrBelhiti** [comment]
Had this same issue in React.

### Relevant package versions:
**@azure/msal-browser: ^2.32.2**
**@azure/msal-react: ^1.5.2**
**react: ^18.2.0**
**react-router-dom: ^6.3.0**

### Possible Root Cause
I suspect that there's a race condition between mzal's redirect handler and the web app's custom redirect handler. mzal's handler should complete execution before the custom handler. However, there is no way to ensure this with the current flow. As a result, the custom handler may observe an inconsistent auth state (e.g. no account object even after user authentication).

### Workaround in React
```
const MicrosoftAuthRedirect = () => {
    const navigate = useNavigate();
    const { instance, accounts } = useMsal();
    const dispatch: Dispatch<AnyAction> = useAppDispatch();

    /* Component making auth request sets desired redirect URL in session storage */
    const redirectUrl: string | null = window.sessionStorage.getItem("redirect-url");

    /* Request to our custom backend which exchanges Microsoft access token to our own access token */
    const [microsoftLogin] = useMicrosoftLoginMutation();

    /* Called upon successfully receiving access token issued by our backend */
    const onLogin = (result: LoginSuccessResponse) => {
        dispatch(setToken({ token: result.accessToken }))
        navigate(redirectUrl || "/home");
    }

    /* Called upon successfully receiving access token issued by Microsoft */
    const onResponse = (value: AuthenticationResult) => {
        microsoftLogin({ accessToken: value.accessToken }).unwrap().then(result => onLogin(result), error => console.error(error));
    }

    /* PRECONDITION: Account object exists (i.e. accounts.length > 0)
       Called when account object is present. DO NOT immediately call
       on page load since this will result in a race condition with mzal's
       own redirect handler. Instead, utilize the useEffect(..., [accounts])
       hook to ensure mzal's redirect handler has populated the account object
       before attempting to access it.
    */
    const requestToken = () => {
        instance.acquireTokenSilent({
            ...loginRequest,
            account: accounts[0]
        }).then(result => {
            onResponse(result);
        });
    }

    /* This was the fix to the race condition */
    useEffect(() => {
        if (accounts.length > 0) {
            requestToken();
        }
    }, [accounts]);

    return (
        <h1>Redirecting...</h1>
    );
}
```
